Use loop and respect the limit parameter for seL4_TCB_WriteRegisters and seL4_TCB_ReadRegisters

ZhanYF opened the pull request:
Use loop and respect the limit parameter for seL4_TCB_WriteRegisters and seL4_TCB_ReadRegisters.

This addresses issue #1085 and currently only affects arm64.

Diff for the generated sel4_client.h:
https://gist.github.com/ZhanYF/3b36cb35591352bd1d094b1724ef7993

I tested this with libvmm on Odroidc4 and qemu and it worked OK, please let me know if you have any feedback.

Indanz reviewed (commented):
Thanks for spending time on this.

Although your solution solves the initial issue for Aarch64, the problem applies to all architectures. I would be fine with writing the few first registers unnecessarily to get rid of the magic 2 and 4 numbers.

It only solves it for the syscall wrappers, it doesn't make things any easier for user space, which has the same problem of being forced to set each individual register one by one. Basically, the user space API is horrible, but we also don't want to break it.

Normally the solution for that would be to use an anonymous union. That would mean changing `seL4_UserContext`, wrap all existing fields in an unnamed union and add an union containing a `regs[]` array of the right size. Doing this is standard C11 and `seL4_UserContext` is only used by user space, so I think it should be fine.

Indanz commented:
Surely that's a long-term project that will take years. By then we can remove the union and only keep the array. Or we just define two structures and add a wrapper that calls the verified one with a cast (and I'd suggest making the array version verified, as that is closer to the actual kernel code).

lsf37 commented:
> Surely that's a long-term project that will take years. 

That doesn't mean we should let the code diverge from the verification subset. We've just spent extra work last year to bring it into the subset again. There is a solution without unions, I see no reason to not use that.

ZhanYF commented:
Thanks for your feedback! Would it be an acceptable solution if I port my changeset to other archs and get rid of the magic 2 and 4 numbers? I think the userspace API is still an issue but that can be addressed at a later stage.

Indanz commented:
When #1551 gets merged it may be easier to add a new, duplicate function that uses arrays.

ZhanYF committed:
Respect count in TCB_Write/ReadRegisters via loop

Previously TCB_WriteRegisters/ReadRegisters ignored the count parameter
and unconditionally copied the entire seL4_UserContext, this leads to
unnecessary overhead for VMM operations.

This addresses issue #1085

Signed-off-by: Yifei Zhan <yifei@zhan.science>

ZhanYF commented:
@Indanz 

Hi Indan, I have reworked this PR to make it work on other architectures too, regarding to magic number, the offset of 2 is still needed to avoid override resume_target and count, as part of the definition for WriteRegister. I have documented why it's needed, so hopefully it's no longer confusing.

Does this look good to you?

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -747,15 +763,27 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
+                    result.append("\tif (count > %d) {" % num_mrs)
+                    result.append("\t\tfor (seL4_Word i=0; i<count-%d; i++) {" % num_mrs)
+                    result.append("\t\t\t((seL4_Word *) &%s->%s)[i] = seL4_GetMR(i+%d);" %
+                                  (param.name, members[num_mrs], num_mrs))
```
What about getting rid of the first if-check by making it part of the for-loop conditional:
```suggestion
                    result.append("\tfor (seL4_Word i=%d; i < count; i++) {" % num_mrs)
                    result.append("\t\t((seL4_Word*)&%s->%s)[i] = seL4_GetMR(i);" %
                                  param.name, members[0]))
```

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -691,17 +691,33 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
-        # Initialise buffered parameters
-        for i in range(num_mrs, len(input_expressions)):
-            result.append("\tseL4_SetMR(%d, %s);" % (i, input_expressions[i]))
-        result.append("")
```
Why did you remove the last `result.append("")`?

I think you may want to keep it, it's done elsewhere too for some reason. But put it after your new for-loop.

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -691,17 +691,33 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
+          result.append("\tif (count > %d) {" % regs_by_mrs)
+          result.append("\t\tfor (seL4_Word i=0; i<count-%d; i++) {" % regs_by_mrs)
+          result.append("\t\t\tseL4_SetMR(i+%d, ((seL4_Word *) &%s)[i]);" %
+                        (num_mrs, input_expressions[num_mrs]))
```
What about getting rid of the first if-check by making it part of the for-loop conditional:
```suggestion
          regs_by_mrs = num_mrs - 2
          result.append("\tfor (seL4_Word i = %d; i < count; i++) {" % regs_by_mrs)
          result.append("\t\tseL4_SetMR(i + 2, ((seL4_Word*)&%s)[i]);" %
                        input_expressions[0])
```

Indanz commented:
Over all looks good to me, much better to only have the magic 2 offset into the MRs which is indeed unavoidable, because it's the number of arguments used for other syscall parameters. With `regs_by_mrs` it's all fine now I think.

Indanz commented:
Could you post an updated diff too please?

Can you fix the CI errors? For Gitlint we currently get:

```
Checking the following commits:
bbe8a0ed Use loop and respect the limit parameter for seL4_TCB_WriteRegisters/ReadRegisters.

1: T1 Title exceeds max length (83>50): "Use loop and respect the limit parameter for seL4_TCB_WriteRegisters/ReadRegisters."
1: T3 Title has trailing punctuation (.): "Use loop and respect the limit parameter for seL4_TCB_WriteRegisters/ReadRegisters."
3: B1 Line exceeds max length (182>72): "Previously TCB_WriteRegisters/ReadRegisters ignored the count parameter and unconditionally copied the entire seL4_UserContext, this leads to unnecessary overhead for VMM operations."
Error: Action gitlint failed.
```

and the style complaints about the indentation being off.

ZhanYF commented:
@Indanz 

Thanks for your feedbacks! Change from the last version:

- Merged the first if-condition with the for-loop.
- There is only one message register for ia32-mcs, to not break it I added a condition check to bypass if num_mrs < 2.
- Check interface_name == "seL4_TCB", otherwise it changes seL4_X86_VCPU_WriteRegisters on x86, which does not take a `count` parameter, we don't any test for it yet and I only noticed it when manually reviewing the generated diff. I think the next step would be
https://github.com/seL4/sel4test/issues/115
- New sel4_client.h diff for arm64 and x86_64: https://gist.github.com/ZhanYF/76fcdaf9264fc7fba8cbc243ba0b03d3
- Fixed style/gitlint CI (Hopefully it will pass, GitHub CI doesn't allow me to reliablly check from my side)

Indanz commented:
Thanks!

>     * There is only one message register for ia32-mcs, to not break it I added a condition check to bypass if num_mrs < 2.

Great, I didn't know that. Let me see if I can fix that, because that's ridiculous.

> I think the next step would be
>       [Add tests for vCPU objects sel4test#115](https://github.com/seL4/sel4test/issues/115)

Yes, this is on my TODO list. However, this will be a lot of work because we don't want to run a real OS, but have our own test code as VM to do specific tests.

>     * New sel4_client.h diff for arm64 and x86_64: https://gist.github.com/ZhanYF/76fcdaf9264fc7fba8cbc243ba0b03d3

Thanks, that looks fine.

>     * Fixed style/gitlint CI (Hopefully it will pass, GitHub CI doesn't allow me to reliablly check from my side)

You can run the `tools/seL4/misc/style*` scripts locally too. They do modify the code, so make sure you either committed or staged all your own changes first before running them.

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -691,17 +691,34 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
+            # Initialise in-register parameters
+            for i in range(num_mrs):
+                if i < len(input_expressions):
+                    result.append("\tmr%d = %s;" % (i, input_expressions[i]))
```
I liked your previous version where you didn't duplicate this bit better, like you do for ReadRegisters.

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -691,17 +691,34 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
-        result.append("")
+        if (method_name == "WriteRegisters" and
+            interface_name == "seL4_TCB" and
+                num_mrs >= 2):
```
If you create a local var to hold this expression, it's out of the way and you can use the shorter variable new instead below.

Indanz commented on libsel4/tools/syscall_stub_gen.py:
```diff
@@ -747,15 +764,25 @@ def generate_stub(arch, wordsize, interface_name, method_name, method_id, input_
+                        result.append("\t%s->%s = %s;" %
+                                      (param.name, members[i], words[i] % source_words))
+                    result.append("\tfor (seL4_Word i=%d; i<count; i++) {" % num_mrs)
+                    result.append("\t\t((seL4_Word *) &%s->%s)[i-%d] = seL4_GetMR(i);" %
```
```suggestion
                    result.append("\tfor (seL4_Word i = %d; i < count; i++) {" % num_mrs)
                    result.append("\t\t((seL4_Word *) &%s->%s)[i - %d] = seL4_GetMR(i);" %
```
Please add some space.

ZhanYF commented:
@Indanz 

Thanks for the new feedbacks, changes from the last version:
- removed duplicate logic in the if-section
- added some spaces to the generated C code
- moved loop check conditions into a local var
- new diff for x86/aarch64 (only style changes since the last version, as expected)
- https://gist.github.com/ZhanYF/0c3f93353329157c05df0999bc8e8530

Indanz reviewed (approved):
Good enough.

Indanz committed:
Merge branch 'master' into yifei/optimization-1085

Indanz commented:
Thanks for your patience and dedication getting this merged!
